Do not suspend DPC testing when clocks are not synchronized

milan-zededa opened the pull request:
If DPC is failing connectivity tests (due to some transient network issue) and device is powered off for long enough that clocks loose time (or there is no battery to keep them running), device will not be able to establish connectivity after power on for a very long time. This is because we demand that failed DPC is tried again only after some time elapses (5 minutes) since the last failed connectivity test. However, when device starts with clocks reset at time "zero" (i.e. start of the epoch), it will take many decades until the current time (as reported by clocks) is past the last failure timestamp. The solution is to detect unsync clocks and allow to test and use DPC immediately.

rene commented:
@eriknordmark , this issue was observed in many ARM devices, which don't have a battery packed RTC. This patch was tested and it solves the issue described in the PR.

rene commented:
@milan-zededa , it seems you will need to adjust the unit tests as well...

eriknordmark reviewed (approved):
This looks fine, but what is the behavior without this fix? Does nim not configure any interfaces hence ntp will never synchronize?
Or does it produce noise and delay but it does synchronize at some point in time?

If the former, is this something we need to backport to stable releases?

eriknordmark commented:
Looks like unit test needs to be updated.

eriknordmark commented:
Apart from this local fix I think we need to look at ntp and time synchronization more broadly.
There might be cases where the DHCP server does not return a NTP server and some firewall blocks access to pool.ntp.org which means the device is running with its free running clock. (For that I think we need to check if ntpd has synchronized and if not do some fallback like get some approximate time from the controller.)

rene commented:
> This looks fine, but what is the behavior without this fix? Does nim not configure any interfaces hence ntp will never synchronize? Or does it produce noise and delay but it does synchronize at some point in time?
> 
> If the former, is this something we need to backport to stable releases?

@eriknordmark , it does not configure interfaces at all. Yes, I think we need to backport this fix to stable releases.

milan-zededa commented:
I have fixed the unit tests. Actually, I rewrote them completely because they looked to be obsolete already before my change (not really testing the current behavior).
I have also added `stable` tag. I agree that we should add this fix to stable releases.
@eriknordmark I agreee that we need more robust approach to time synchronization. This quick fix will unblock this particular customer but I will also add story for a more broad solution.

eriknordmark reviewed (approved):
LGTM
